numpyro/infer/hmc_gibbs.py

```diff
@@ -149,3 +153,178 @@ def potential_fn(z_gibbs, z_hmc):
+    :class:`HMCGibbs` will be marginalized out by default (if they have enumerate supports).
+
+    **References**
+
+    1. *Peskun's theorem and a modified discrete-state Gibbs sampler*,
+       Liu, Jun S. (1996)
+
+    **Example**
+
+    .. doctest::
+
+        >>> from jax import random
+        >>> import jax.numpy as jnp
+        >>> import numpyro
+        >>> import numpyro.distributions as dist
+        >>> from numpyro.infer import MCMC, NUTS, HMCGibbs, discrete_gibbs_fn
+        ...
+        >>> def model(probs, locs):
+        ...     c = numpyro.sample("c", dist.Categorical(probs))
+        ...     numpyro.sample("x", dist.Normal(locs[c], 0.5))
+        ...
+        >>> probs = jnp.array([0.15, 0.3, 0.3, 0.25])
+        >>> locs = jnp.array([-2, 0, 2, 4])
+        >>> gibbs_fn = discrete_gibbs_fn(model, probs, locs)
+        >>> kernel = HMCGibbs(NUTS(model), gibbs_fn, gibbs_sites=["c"])
+        >>> mcmc = MCMC(kernel, 1000, 100000, progress_bar=False)
+        >>> mcmc.run(random.PRNGKey(0), probs, locs)
+        >>> samples = mcmc.get_samples()["x"]
+        >>> assert abs(jnp.mean(samples) - 1.3) < 0.1
+
+    """
+    # NB: all of the information such as `model`, `model_args`, `model_kwargs`
+    # can be accessed from HMCGibbs.sample but we require them here to
+    # simplify the api of `gibbs_fn`
+    prototype_trace = trace(seed(model, rng_seed=0)).get_trace(*model_args, **model_kwargs)
+    support_sizes = {
+        name: jnp.broadcast_to(site["fn"].enumerate_support(False).shape[0], jnp.shape(site["value"]))
+        for name, site in prototype_trace.items()
+        if site["type"] == "sample" and site["fn"].has_enumerate_support and not site["is_observed"]
+    }
+
+    def gibbs_fn(rng_key, gibbs_sites, hmc_sites):
+        # convert to unconstrained values
+        z_hmc = {k: biject_to(prototype_trace[k]["fn"].support).inv(v)
+                 for k, v in hmc_sites.items() if k in prototype_trace}
+        enum = any(site["type"] == "sample"
+                   and not site["is_observed"]
+                   and site["fn"].has_enumerate_support
+                   and name not in gibbs_sites
+                   for name, site in prototype_trace.items())
+
+        def potential_fn(z_discrete):
+            model_kwargs["_gibbs_sites"] = z_discrete
+            return potential_energy(_wrap_model(model), model_args, model_kwargs, z_hmc, enum=enum)
+
+        # get support_sizes of gibbs_sites
+        support_sizes_flat, _ = ravel_pytree({k: support_sizes[k] for k in gibbs_sites})
+        num_discretes = support_sizes_flat.shape[0]
+
+        rng_key, rng_permute = random.split(rng_key)
+        idxs = random.permutation(rng_key, jnp.arange(num_discretes))
+
+        def body_fn(i, val):
+            idx, support_size = idxs[i], support_sizes_flat[i]
+            return _discrete_step(*val, potential_fn=potential_fn, idx=idx, support_size=support_size)
```

Comment:
this does many MH corrections instead of one. is that preferred?

Reply:
This is more like a sequence of MH updates in a Gibbs scheme (var1 -> var2 -> var3 -> ...). I think I should change the implementation to the Metropolized/Modified Gibbs sampler in reference [1], that is
+ propose a discrete sample proportional to its conditional distribution (this requires evaluating log_density for all possible values in the support)
+ modify accept prob to `(1 - p_old) / (1 - p_new)`
and rename the current implementation to discrete_metropolis_fn. What do you think?

Reply:
Reading literatures through the day, I think it is better to use Gibbs update, instead of metropolized version here. The motivation of metropolized sampler is to try to move away from the current position. But assuming that the discrete variable is Bernoulli with prob=0.5, using metropolized version will always give acceptance probably = 1, hence the sampled sequence will be something like 0, 1, 0, 1, 0, 1,... which is periodic. I think I should use Gibbs sampler from now on, even in mixed hmc implementation. (for mixed hmc, when calculating Delta_E, we should try to remove the involvement of Q...)